File: Test Case Generation/deepseek/cpp/test_task_6.cpp
#include <gtest/gtest.h>
#include <vector>
#include <string>
#include "parse_nested_parens.h" // Assuming the function is declared here

using namespace std;

// Test fixture for parse_nested_parens function
class ParseNestedParensTest : public ::testing::Test {
protected:
    void SetUp() override {}
    void TearDown() override {}
};

// Test empty string
TEST_F(ParseNestedParensTest, EmptyString) {
    vector<int> result = parse_nested_parens("");
    EXPECT_TRUE(result.empty());
}

// Test single group with no nesting
TEST_F(ParseNestedParensTest, SingleGroupNoNesting) {
    vector<int> result = parse_nested_parens("()");
    vector<int> expected = {1};
    EXPECT_EQ(result, expected);
}

// Test single group with nesting level 2
TEST_F(ParseNestedParensTest, SingleGroupNesting2) {
    vector<int> result = parse_nested_parens("(())");
    vector<int> expected = {2};
    EXPECT_EQ(result, expected);
}

// Test single group with nesting level 3
TEST_F(ParseNestedParensTest, SingleGroupNesting3) {
    vector<int> result = parse_nested_parens("((()))");
    vector<int> expected = {3};
    EXPECT_EQ(result, expected);
}

// Test single group with multiple parentheses but max nesting level 2
TEST_F(ParseNestedParensTest, SingleGroupMultiplePairsNesting2) {
    vector<int> result = parse_nested_parens("(()())");
    vector<int> expected = {2};
    EXPECT_EQ(result, expected);
}

// Test multiple groups from example
TEST_F(ParseNestedParensTest, MultipleGroupsExample) {
    vector<int> result = parse_nested_parens("(()()) ((())) () ((())()())");
    vector<int> expected = {2, 3, 1, 3};
    EXPECT_EQ(result, expected);
}

// Test groups separated by multiple spaces
TEST_F(ParseNestedParensTest, MultipleSpacesBetweenGroups) {
    vector<int> result = parse_nested_parens("()   (())    ((()))");
    vector<int> expected = {1, 2, 3};
    EXPECT_EQ(result, expected);
}

// Test leading and trailing spaces
TEST_F(ParseNestedParensTest, LeadingAndTrailingSpaces) {
    vector<int> result = parse_nested_parens("  (())  ()  ");
    vector<int> expected = {2, 1};
    EXPECT_EQ(result, expected);
}

// Test complex nesting patterns
TEST_F(ParseNestedParensTest, ComplexNesting) {
    vector<int> result = parse_nested_parens("((()())) (()(()))");
    vector<int> expected = {3, 3};
    EXPECT_EQ(result, expected);
}

// Test single group with maximum nesting at beginning
TEST_F(ParseNestedParensTest, MaxNestingAtBeginning) {
    vector<int> result = parse_nested_parens("((()))()");
    vector<int> expected = {3, 1};
    EXPECT_EQ(result, expected);
}

// Test single group with maximum nesting in middle
TEST_F(ParseNestedParensTest, MaxNestingInMiddle) {
    vector<int> result = parse_nested_parens("(()(()))");
    vector<int> expected = {3};
    EXPECT_EQ(result, expected);
}

// Test groups with varying levels
TEST_F(ParseNestedParensTest, VaryingLevels) {
    vector<int> result = parse_nested_parens("() ((())) (()) ()");
    vector<int> expected = {1, 3, 2, 1};
    EXPECT_EQ(result, expected);
}

// Test deep nesting
TEST_F(ParseNestedParensTest, DeepNesting) {
    vector<int> result = parse_nested_parens("((((()))))");
    vector<int> expected = {5};
    EXPECT_EQ(result, expected);
}

// Test unbalanced parentheses - edge case (assuming valid input as per spec)
// Note: The function assumes valid input, but testing gracefully handles it
TEST_F(ParseNestedParensTest, UnbalancedParenthesesExtraOpen) {
    // The function will still return a result based on its logic
    // Groups are only pushed when level returns to 0
    vector<int> result = parse_nested_parens("(()");
    // Since level never returns to 0, no groups are pushed
    EXPECT_TRUE(result.empty());
}

TEST_F(ParseNestedParensTest, UnbalancedParenthesesExtraClose) {
    // Starting with close parenthesis causes level to go negative
    // But we still process until a group completes or end of string
    vector<int> result = parse_nested_parens("())");
    // Level starts at 0, sees ')', becomes -1, then sees '(' etc.
    // The first group ends when level returns to -1? Actually level becomes -2 etc.
    // For "())", sequence: ')'(level=-1), ')'(level=-2), never reaches 0 from positive
    // So no groups pushed
    EXPECT_TRUE(result.empty());
}

// Test single character (no parentheses)
TEST_F(ParseNestedParensTest, NoParentheses) {
    vector<int> result = parse_nested_parens("abc");
    EXPECT_TRUE(result.empty());
}

// Test spaces only
TEST_F(ParseNestedParensTest, SpacesOnly) {
    vector<int> result = parse_nested_parens("     ");
    EXPECT_TRUE(result.empty());
}

// Test groups with no spaces between them (invalid per spec but testing code behavior)
TEST_F(ParseNestedParensTest, NoSpacesBetweenGroups) {
    vector<int> result = parse_nested_parens("()()()");
    vector<int> expected = {1, 1, 1};
    EXPECT_EQ(result, expected);
}

// Test empty group (impossible with valid parentheses, but for completeness)
TEST_F(ParseNestedParensTest, EmptyParenthesesGroup) {
    vector<int> result = parse_nested_parens("()");
    vector<int> expected = {1};
    EXPECT_EQ(result, expected);
}

// Test large number of groups
TEST_F(ParseNestedParensTest, ManyGroups) {
    string input;
    for (int i = 0; i < 100; i++) {
        input += "() ";
    }
    input.pop_back(); // Remove trailing space
    
    vector<int> result = parse_nested_parens(input);
    EXPECT_EQ(result.size(), 100);
    for (int level : result) {
        EXPECT_EQ(level, 1);
    }
}

// Test mixed complexity
TEST_F(ParseNestedParensTest, MixedComplexity) {
    vector<int> result = parse_nested_parens("((())) (()()) ((())()) (((())))");
    vector<int> expected = {3, 2, 3, 4};
    EXPECT_EQ(result, expected);
}

// Test string with tabs instead of spaces
TEST_F(ParseNestedParensTest, TabsAsSeparators) {
    // The function only checks spaces, so tabs won't separate groups
    // This tests the actual behavior
    vector<int> result = parse_nested_parens("()\t(())");
    // Since '\t' is not space, everything is treated as one group
    vector<int> expected = {2}; // "()(())" has max nesting 2
    EXPECT_EQ(result, expected);
}

// Test random spaces
TEST_F(ParseNestedParensTest, RandomSpaces) {
    vector<int> result = parse_nested_parens("  (())   ()   ((()))  ");
    vector<int> expected = {2, 1, 3};
    EXPECT_EQ(result, expected);
}

// Test with newlines and spaces (assuming input doesn't have newlines per spec)
TEST_F(ParseNestedParensTest, NewlinesAndSpaces) {
    // Newlines are not spaces, so they'd be part of the group
    vector<int> result = parse_nested_parens("()\n()");
    // Treated as single group "()\n()" which is invalid parentheses
    // So no groups completed if newline doesn't close properly
    // The function only looks for '(', ')', and spaces
    // Newline will be ignored as it's not recognized
    // Actually, the code processes each char, ignoring those not '(' or ')'
    // So "()\n()" becomes two groups because '\n' is not space? Wait, the code only checks for spaces specifically
    // '\n' is not space, so the whole string is one continuous stream: '(',')', '\n', '(', ')'
    // Since '\n' is not '(' or ')', it's just skipped. So we have "( ) ( )" -> "()()" which forms two groups
    // Let me trace: 
    // char[0]='(', level=1,max=1
    // char[1]=')', level=0 -> push(1)
    // char[2]='\n', no change
    // char[3]='(', level=1,max=1
    // char[4]=')', level=0 -> push(1)
    // Result: {1,1}
    vector<int> expected = {1, 1};
    EXPECT_EQ(result, expected);
}

File: Test Case Generation/deepseek/python/test_task_6.py
import pytest
from typing import List
from your_module import parse_nested_parens

class TestParseNestedParens:
    """Test suite for parse_nested_parens function"""
    
    def test_basic_example(self):
        """Test the example from the docstring"""
        result = parse_nested_parens('(()()) ((())) () ((())()())')
        assert result == [2, 3, 1, 3]
    
    def test_single_group_simple(self):
        """Test a single group with simple parentheses"""
        assert parse_nested_parens('()') == [1]
        assert parse_nested_parens('(())') == [2]
        assert parse_nested_parens('((()))') == [3]
    
    def test_single_group_complex(self):
        """Test a single group with varying nesting levels"""
        assert parse_nested_parens('(()())') == [2]
        assert parse_nested_parens('((())())') == [3]
        assert parse_nested_parens('(()(()))') == [3]
    
    def test_multiple_groups(self):
        """Test multiple groups with spaces"""
        assert parse_nested_parens('() ()') == [1, 1]
        assert parse_nested_parens('(()) ()') == [2, 1]
        assert parse_nested_parens('((())) (())') == [3, 2]
        assert parse_nested_parens('(()()) ((()))') == [2, 3]
    
    def test_empty_string(self):
        """Test empty string input"""
        assert parse_nested_parens('') == []
    
    def test_string_with_only_spaces(self):
        """Test string containing only spaces"""
        assert parse_nested_parens('   ') == []
        assert parse_nested_parens(' ') == []
    
    def test_trailing_and_leading_spaces(self):
        """Test string with leading and trailing spaces"""
        assert parse_nested_parens('  ()  ') == [1]
        assert parse_nested_parens('  (())  ()  ') == [2, 1]
    
    def test_multiple_consecutive_spaces(self):
        """Test string with multiple spaces between groups"""
        assert parse_nested_parens('()   ()') == [1, 1]
        assert parse_nested_parens('(())     ((()))') == [2, 3]
    
    def test_complex_nested_patterns(self):
        """Test complex nested parentheses patterns"""
        assert parse_nested_parens('((())(()))') == [3]
        assert parse_nested_parens('(()()(()))') == [3]
        assert parse_nested_parens('((((()))))') == [5]
    
    def test_multiple_complex_groups(self):
        """Test multiple complex groups together"""
        assert parse_nested_parens('((())) (()()) ((())())') == [3, 2, 3]
        assert parse_nested_parens('(((()))) (()(())) ()') == [4, 3, 1]
    
    def test_deep_nesting(self):
        """Test very deep nesting levels"""
        deep_string = '(' * 100 + ')' * 100
        assert parse_nested_parens(deep_string) == [100]
        
        deep_strings = ' '.join(['(' * i + ')' * i for i in [10, 20, 30, 40]])
        expected = [10, 20, 30, 40]
        assert parse_nested_parens(deep_strings) == expected
    
    def test_mixed_order_groups(self):
        """Test groups with different nesting orders"""
        assert parse_nested_parens('(()()) ((())) () ((())()())') == [2, 3, 1, 3]
        assert parse_nested_parens('((())) () (()()) ((())()())') == [3, 1, 2, 3]
    
    def test_alternating_patterns(self):
        """Test alternating parentheses patterns"""
        assert parse_nested_parens('()(())') == [2]  # Should be parsed as one group
        assert parse_nested_parens('(()())') == [2]
        assert parse_nested_parens('(())()') == [2]  # One continuous group
    
    def test_single_parenthesis_malformed(self):
        """Test malformed input with unbalanced parentheses (function handles anyway)"""
        # The function doesn't validate balanced parentheses, just counts nesting
        assert parse_nested_parens('(') == [1]
        assert parse_nested_parens(')(') == [0]  # depth becomes negative
        assert parse_nested_parens('(()') == [2]  # Unclosed
    
    def test_mixed_malformed_with_valid(self):
        """Test mixture of malformed and valid groups"""
        assert parse_nested_parens('( (()) )') == [2, 2, 1]  # Bare parentheses treated as groups
        assert parse_nested_parens('(() ())') == [2, 1, 1]
    
    def test_no_parentheses(self):
        """Test groups with no parentheses"""
        assert parse_nested_parens('abc def') == [0, 0]
        assert parse_nested_parens('no parentheses here') == [0, 0, 0, 0]
        assert parse_nested_parens('') == []
    
    def test_mixed_characters(self):
        """Test strings with mixed characters (though function ignores non-parens)"""
        assert parse_nested_parens('(a(b)c)') == [2]
        assert parse_nested_parens('(())abc(())') == [2]  # Treated as one group
        assert parse_nested_parens('a (()) b () c') == [0, 2, 0, 1, 0]
    
    def test_whitespace_only_after_split(self):
        """Test that empty strings from split are filtered out"""
        # The function uses 'if x' to filter empty strings
        result = parse_nested_parens('()   ')
        assert result == [1]
        
        result = parse_nested_parens('   ()   ')
        assert result == [1]
    
    def test_edge_case_single_character(self):
        """Test single character inputs"""
        assert parse_nested_parens('(') == [1]
        assert parse_nested_parens(')') == [0]
        assert parse_nested_parens('a') == [0]
        assert parse_nested_parens(' ') == []
    
    def test_performance_large_input(self):
        """Test with large input to ensure function handles it"""
        groups = ['(' * i + ')' * i for i in range(1, 101)]
        large_string = ' '.join(groups)
        result = parse_nested_parens(large_string)
        expected = list(range(1, 101))
        assert result == expected

if __name__ == '__main__':
    pytest.main([__file__, '-v', '--cov=.'])